akka-http-core/src/main/scala/akka/http/impl/engine/http2/HttpMessageRendering.scala

```diff
@@ -75,9 +75,15 @@ private[http2] sealed abstract class MessageRendering[R <: HttpMessage] extends
     HttpMessageRendering.addContentHeaders(headerPairs, r.entity)
     HttpMessageRendering.renderHeaders(r.headers, headerPairs, peerIdHeader, log, isServer = r.isResponse)
 
-    val headersFrame = ParsedHeadersFrame(nextStreamId(r), endStream = r.entity.isKnownEmpty, headerPairs.result(), None)
+    val streamId = nextStreamId(r)
+    val headersFrame = ParsedHeadersFrame(streamId, endStream = r.entity.isKnownEmpty, headerPairs.result(), None)
+    val trailingHeadersFrame: Option[ParsedHeadersFrame] =
+      r.attribute(AttributeKeys.trailer) match {
+        case Some(trailer) => Some(ParsedHeadersFrame(streamId, true, trailer.headers.map(header => (header.name, header.value)), None))
```

Comment:
Since we're converting to pairs here anyway, perhaps we should store the headers as pairs already in `Trailer`, and allow creating the `Trailer` without going through `HttpHeader`. Since we're not currently providing access to the trailing headers for incoming messages anyway, perhaps it's OK to not have public API's to read the headers from the trailer for now anyway.

Reply:
Yes, maybe let's strip down the public API completely for now so that you can only create a `Trailer` from `Seq[HttpHeader]` or pairs but not access the values. So, we can choose whatever representation we want in the implementation.